src/runtime/tracing.cpp

```diff
@@ -10,7 +10,104 @@ typedef int32_t (*trace_fn)(void *, const halide_trace_event_t *);
 
 namespace Halide { namespace Runtime { namespace Internal {
 
-WEAK int halide_trace_file = 0;
+// A spinlock that allows for shared and exclusive access. It's
+// equivalent to a reader-writer lock, but in my case the "readers"
+// will actually be writing simultaneously to the trace buffer, so
+// that's a bad name.
+class SharedExclusiveSpinLock {
+    uint32_t lock;
+    const static uint32_t exclusive_held_mask = 0x80000000;
+    const static uint32_t exclusive_waiting_mask = 0x40000000;
+    const static uint32_t shared_mask = 0x3fffffff;
+
+public:
+    void acquire_shared() {
+        while (1) {
+            uint32_t x = lock & shared_mask;
+            if (__sync_bool_compare_and_swap(&lock, x, x + 1)) {
```

Comment:
Are these intrinsics portable everywhere? Could we use std::atomic instead?

Reply:
We use these elsewhere in the runtime for this sort of synchronization. It's always compiled by clang, so it should be fine.

Reply:
Might add that as a comment so future versions of me won't ask the same q and/or attempt to change them

Reply:
done